### PR TITLE
Http2 upgrade v7

### PR DESCRIPTION
**Pull request by catenacyber**
Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) ticket:
https://redmine.openinfosecfoundation.org/issues/3866

Describe changes:
- Adds mechanism to change protocol from HTTP1 to HTTP2
- Retry protocol detection on protocol change if we did not have enough data
- Fix DNP3 unit tests when compiling with `-DFUZZING_BUILD_MODE_UNSAFE_FOR_PRODUCTION`

[PRScript](https://redmine.openinfosecfoundation.org/projects/suricata/wiki/PRscript) output (if applicable):

suricata-verify-pr: 326

Makes https://github.com/OISF/suricata-verify/pull/326 pass

Modifies #5379 by rebasing to get CI happy

The downside compared to #5354 is to have one more test ` FlowChangeProto(f)` to run at (almost) each call of `AppLayerParserParse`

### CODE REVIEW

File: src/app-layer-htp.c [hunk 1]
```diff
@@ -6406,357 +6406,6 @@ static int HTPParserTest16(void)
     return result;
 }
 
-/** \test CONNECT with plain text HTTP being tunneled */
```
Comment (victorjulien) [inline src/app-layer-htp.c:6409]: Can you put these removals in their own commit?

Comment (catenacyber) [reply src/app-layer-htp.c:6409]: Sure
